src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BedSearchService.kt

```diff
@@ -0,0 +1,134 @@
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.repository.TemporaryAccommodationBedSearchResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.AuthorisableActionResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.ValidatableActionResult
+import java.time.LocalDate
+
+@Service
+class BedSearchService(
+  private val bedSearchRepository: BedSearchRepository,
+  private val postcodeDistrictRepository: PostcodeDistrictRepository,
+  private val characteristicService: CharacteristicService,
+) {
+  fun findApprovedPremisesBeds(
+    user: UserEntity,
+    postcodeDistrictOutcode: String,
+    maxDistanceMiles: Int,
+    startDate: LocalDate,
+    durationInDays: Int,
+    requiredPremisesCharacteristics: List<String>,
+    requiredRoomCharacteristics: List<String>,
+  ): AuthorisableActionResult<ValidatableActionResult<List<ApprovedPremisesBedSearchResult>>> {
+    if (!user.hasRole(UserRole.MATCHER)) {
+      return AuthorisableActionResult.Unauthorised()
+    }
+
+    return AuthorisableActionResult.Success(
+      validated {
+        val premisesCharacteristicErrors = mutableListOf<String>()
+        val premisesCharacteristics = requiredPremisesCharacteristics.mapNotNull {
+          val characteristic = characteristicService.getCharacteristicByPropertyName(it)
+
+          if (characteristic == null) {
+            premisesCharacteristicErrors += "$it doesNotExist"
+            return@mapNotNull null
+          }
+
+          if (!characteristic.matches(ServiceName.approvedPremises.value, "premises")) {
+            premisesCharacteristicErrors += "$it scopeInvalid"
+            return@mapNotNull null
+          }
+
+          characteristic
+        }
+
+        if (premisesCharacteristicErrors.any()) {
+          "$.requiredPremisesCharacteristics" hasValidationError premisesCharacteristicErrors.joinToString(", ")
+        }
+
+        val roomCharacteristicErrors = mutableListOf<String>()
+        val roomCharacteristics = requiredRoomCharacteristics.mapNotNull {
+          val characteristic = characteristicService.getCharacteristicByPropertyName(it)
+
+          if (characteristic == null) {
+            roomCharacteristicErrors += "$it doesNotExist"
+            return@mapNotNull null
+          }
+
+          if (!characteristic.matches(ServiceName.approvedPremises.value, "room")) {
+            roomCharacteristicErrors += "$it scopeInvalid"
+            return@mapNotNull null
+          }
+
+          characteristic
+        }
+
+        if (roomCharacteristicErrors.any()) {
+          "$.requiredRoomCharacteristics" hasValidationError roomCharacteristicErrors.joinToString(", ")
+        }
+
+        postcodeDistrictRepository.findByOutcode(postcodeDistrictOutcode)
+          ?: ("$.postcodeDistrictOutcode" hasValidationError "doesNotExist")
+
+        if (durationInDays < 1) {
+          "$.durationInDays" hasValidationError "mustBeAtLeast1"
```

Comment:
Can this be `"$.durationDays`? It's confusing things UI side that this is `duration*In*Days`, but the field name on `BedSearchParameters` is`durationDays`